File: src/main/java/io/github/harishb2k/spark/sql/parser/logical/optimizer/LogicalPlanOptimizer.java
package io.github.harishb2k.spark.sql.parser.logical.optimizer;


import io.github.harishb2k.spark.sql.parser.node.Node;

import java.util.ArrayList;
import java.util.List;

public class LogicalPlanOptimizer {
    private final List<LogicalRule> rules;

    public LogicalPlanOptimizer() {
        this.rules = new ArrayList<>();
        this.rules.add(new WhereClauseCollapseRuleExt());
    }

    public Object optimize(Node node) {
        for (LogicalRule rule : rules) {
            rule.transform(node);
        }
        for (Node n : node.getChildren()) {
            optimize(n);
        }
        return node;
    }
}

File: src/main/java/io/github/harishb2k/spark/sql/parser/logical/optimizer/LogicalPlanOptimizer.scala
package io.github.harishb2k.spark.sql.parser.logical.optimizer

import io.github.harishb2k.spark.sql.parser.node.{LogicalPlan, UnresolvedSingleSelect}

import java.util
import scala.collection.JavaConversions._

class LogicalPlanOptimizer {
  val rules = new util.ArrayList[Rule]
  rules.add(new UnresolvedJoinRule)

  def optimize(logicalPlan: LogicalPlan): LogicalPlan = {
    var lp = logicalPlan;

    // Run all rules in in input
    for (rule <- rules) {

      // Run a rule on input plan
      val afterApply = rule.apply(lp)

      // UnresolvedSingleSelect is the final result
      if (afterApply.isInstanceOf[UnresolvedSingleSelect]) {
        lp = afterApply
      }
    }

    // return updated plan
    lp
  }
}
